### PR TITLE
doc: Add text about Jenkins + Docker

### LINKED ISSUE #1974: Buildserver documentation
Documentation for build server, please describe:

- [x] tutorial for Jenkinsfile for people not knowing Jenkins, Groovy, or Dockerfiles; with hints/references to:
  - [x] how to add a docker-based build agent
  - [x] how to configure the build job in Jenkins (if Jenkins is newly installed)
  - [x] how to identify which docker image should be used in issue reports (so how to write a good issue which allows reproducibility with our docker images)
  - [x] how to get a docker image to reproduce a problem (for devs)
  - [x] how docker containers and Jenkinsfile are connected.
  - [x] how to add a build job (which compiles Elektra)
  - [x] how to add a script to be executed on the build server (which does checks)
- [x] some more comments in Jenkinsfile (about structure, main, ...)
- [x] when which job gets triggered (push to master, PRs, or debian, see also #2027)
- [x] section about deployment in BUILDSERVER.md, see also #2003
- [x] describe which Jenkinsfile is used (if PR and master both make changes to the Jenkinsfile)
- [x] bring scripts/docker/README.md up-to-date
- [x] overview of Dockerfiles in scripts/docker

If some parts of the documentation already exist for this release, it would be great.

### PR DESCRIPTION
**Pull request by ingwinlu**
Resolves #1974

## Basics

Do not describe the purpose here but:

- [ ] Short descriptions should be in the release notes (added as entry in
      doc/news/_preparation_next_release.md which contains `*(my name)*`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed:

@markus2330 please review my pull request

### CODE REVIEW

File: doc/BUILDSERVER.md [hunk 1]
```diff
@@ -196,3 +196,71 @@ as much as possible.
 As for labels `gitmirror` should be if you want to cache repositories on this
 node.
 If Docker is available the `docker` label should be set.
+
+### Understand jenkins output
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: jenkins -> Jenkins

File: doc/BUILDSERVER.md [hunk 2]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+Our jenkins build uses parallel steps inside a single build job to do most of
+the work.
+To reliable determine which stages failed it is best to look over the build
+results in the new Jenkins Blue Ocean view.
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: add link? remove "new".

File: doc/BUILDSERVER.md [hunk 3]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+On selecting a stage you can further expand all steps in the stage.
+Of interest should be one of the first steps which echos out the Docker image
+used to run the stage.
+You also want to look for whatever failed (which should be in a step also marked
+red to indicate failure).
+
+### Reproducing buildserver errors locally
+To reproduce the docker image used during the failing test you have to rebuild
+them locally.
+Determine which image is used as described above.
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: where "above"?

File: doc/BUILDSERVER.md [hunk 4]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+
+Afterwards you can locate the image via the `DOCKER_IMAGES` map in the
+Jenkinsfile.
+It maps the stage that failed to the Dockerfile in the repository.
+Now you can build the image:
+
+```
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: current working directory?

File: doc/BUILDSERVER.md [hunk 5]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+docker build --build-arg JENKINS_GROUPID=`id -g` \
+             --build-arg JENKINS_USERID=`id -u` \
+             -t libelektra-debian-stable \
+             scripts/docker/debian/stretch/
+```
+
+To run the build commands in the container run the following from libelektras
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: libelektra's

File: doc/BUILDSERVER.md [hunk 6]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+```
+docker run -it \
+           -v `pwd`:/home/jenkins/ws \
+           -w /home/jenkins/ws \
+           libelektra-debian-stable
+```
+
+### Add new test stages
+Test stages are determined by the Jenkinsfile used by our build job.
+It currently is located at `scripts/jenkins/Jenkinsfile`.
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: They are currently located at + link to file.

File: doc/BUILDSERVER.md [hunk 7]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+
+First you need to determine where and when it makes sense to run your desired
+test case.
+Keep in mind that resources are limited so you only want to use them if
+absolutly needed.
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: Formatting. Line breaks? Reminder maybe as quote?

File: doc/BUILDSERVER.md [hunk 8]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+Most additional tests run in the 'full' build stages.
+A good approach is to read similar test cases and adopt them to your needs.
+
+There are some helper functions defined in the Jenkinsfile which should make it
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: This text does not help. Which of the functions are helper, which are stages. No example given.

File: doc/BUILDSERVER.md [hunk 9]
```diff
@@ -196,3 +196,71 @@ as much as possible.
...
+
+### Modify test environments
+You can also modify the test environments (update a dependency, install a new
+dependency, ...) by editing the Dockerfiles checked into SCM.
+Determine which ones you need to modify by tracing which images are used for
+what stages via the `DOCKER_IMAGES` map used in our Jenkinsfile.
+The `dockerInit` function should be a good start point.
+
+Following docker best practises you should try to minimize layer count
+when possible.
```
Comment (markus2330) [inline doc/BUILDSERVER.md]: Not finished?

File: doc/BUILDSERVER.md (hunk 7 above)
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: What do you mean with Formatting and Line breaks?

**Comment by ingwinlu** [comment]
I restructured much of BUILDSERVER.md as i noticed that most points raised were already described in different sections.

**Comment by markus2330** [comment]
Sorry, I cannot see any improvement. Not even trivial typos like "Jenkins" are fixed nor release notes were updated.

File: doc/BUILDSERVER.md [hunk 10]
```diff
@@ -32,22 +32,54 @@ or if just the pull request should be checked:
...
+tests called `libelektra`.
+
+Simplified a Multibranch Pipeline Job acts as an umbrella job that spawns
+child jobs depending on branch (hence multibranch).
+The main purpose of the libelektra job is to scan the repository for
+changes to existing branches or to find new ones (for example branches
+that are used in PR's).
+It also contains the information on how to build the jobs in the form
+of a path that points to a Jenkinsfile containing more detailed instructions
+as well as some configuration that should be shared by all resulting jobs.
```
[the remark this answers is not in the document]
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: Mostly internal properties were ment here. I will remove the sentence.

File: doc/BUILDSERVER.md [hunk 11]
```diff
@@ -59,45 +91,18 @@ list of currently available syntax can be seen in
 syntax](https://build.libelektra.org/jenkins/job/libelektra/pipeline-syntax/)
 after a login to the build server.
 
-## Security
+We also provide a number of helper functions in our Jenkinsfiles that are
```
[the remark this answers is not in the document]
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: where -> in our jenkinsfile
how they are called -> too many to list. basically all functions defined in the jenkinsfile could be called helpers in some way.

File: doc/BUILDSERVER.md [hunk 12]
```diff
@@ -196,3 +197,109 @@ as much as possible.
...
+It is the default View opened when accessing the build results from github.
+For libelektra the Url's are
+https://build.libelektra.org/jenkins/job/libelektra/ and
+https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/branches/
+.
+
+Failed stages are marked in red.
+On selecting a stage you can further expand all steps in the stage.
+Of interest should be one of the first steps which echos out the Docker image
+used to run the stage.
```
[the remark this answers is not in the document]
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: Was not possible when I originally wrote it (no nested stages in parallel stages).

File: doc/BUILDSERVER.md (hunk 9 above)
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: the best practise is to minimize layer count. If that is not good enough we can remove the sentence as well.

File: doc/BUILDSERVER.md [hunk 13]
```diff
@@ -196,3 +197,109 @@ as much as possible.
...
+
+- all of the **standard PR jobs** were already **successful**, and
+- you are sure that you **do not want change anything** in your PR anymore
+
+.
+
+# Issues with the build environment
+If you have issues that are related to the build system you can open a normal
+issue and tag it with `build` and `question`.
+If you feel like your inquiry does not warrent a issue on its own, please use
```
[the remark this answers is not in the document]
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: Only moved this section. No idea what the original author ment.

File: doc/BUILDSERVER.md [hunk 14]
```diff
@@ -196,3 +197,109 @@ as much as possible.
...
+[our buildserver issue](https://issues.libelektra.org/160).
```
[the remark this answers is not in the document]
Comment (ingwinlu) [reply doc/BUILDSERVER.md]: Only moved this section.

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**